lib/file-key-info.js

```diff
@@ -1,17 +1,16 @@
-var StringKeyInfo = require("./string-key-info"),
-  fs = require("fs");
+const fs = require("fs");
+const KeyInfo = require("./key-info");
 
 /**
- * A key info provider implementation
+ * A key info provider implementation for certificates in files
  *
- * @param {string} file path to public certificate
+ * @param {string} file path to public certificate, in base64 DER format, not PEM format
  */
-function FileKeyInfo(file) {
-  var key = fs.readFileSync(file);
-  StringKeyInfo.apply(this, [key]);
+class FileKeyInfo extends KeyInfo {
+  constructor(file) {
+    const cert = fs.readFileSync(file);
```

Comment:
```suggestion
      var pem = fs.readFileSync(file, 'utf8');
      var certificates = pem.match(/-----BEGIN CERTIFICATE-----[^-]*-----END CERTIFICATE-----/g);
      super(certificates.map(function(certificate) {
        return certificate.replace(/-----BEGIN CERTIFICATE-----|-----END CERTIFICATE-----|\n|\r/g, '');
      }));
```

Reply:
That breaks tests since some of the certificates are stored in a binary format, so I didn't want to have to change tests or break compatibility with that format of cert.

Reply:
I noticed there are no tests for the KeyInfo class.
Can you make a test where it accepts string/array of string and then sign/validate?

Reply:
We shouldn't have to test it separately as long as we have complete coverage. The code coverage scanning I've added helps us to see that we're pretty covered here. There are lots of tests that do just what you say but tests more parts of the code than just `KeyInfo`. Is there any part of the coverage that you feel is missing?

Honestly, we could probably check to make sure that the `<X509Certificate />` element is there. I don't know if I have time to write that test anytime soon. I've spent more on this than I wanted to. Do you have time to contribute a test @shunkica ?

Reply:
Let's say I want to add the entire certificate chain during the signature process to the KeyInfo element using one of the providers you created.

I have my certificate chain in pem format, and my private key in pem format.

```
    var xml = "<root><x /></root>";
    var sig = new SignedXml();
    sig.signingKey = fs.readFileSync("./test/static/myprivate.pem");
    sig.keyInfoProvider = new FileKeyInfo("./test/static/mypublic_chain.pem");
    sig.computeSignature(xml);
    var signedXml = sig.getSignedXml();
```
This results in:
```
TypeError: Cannot read properties of null (reading 'match')
    at FileKeyInfo.getKeyInfoContent (lib\key-info.js:65:17)
    at SignedXml.getKeyInfo (lib\signed-xml.js:908:33)
    at SignedXml.computeSignature (lib\signed-xml.js:815:24)
```

Reply:
Here are some of my opinions regarding what the the implementation of the default KeyInfo classes should be.

**During signature:**
- the user can load certs in **PEM string** into the `KeyInfo | FileKeyInfo` constructor (it can be a single cert or a chain)
> from crypto.createPrivateKey(key) -> If key is a string or Buffer, format is assumed to be 'pem'
- only `getKeyInfoContent()` used during xml generation - how do we get the certs inside getKeyInfoContent?
  - we already have `this.certs`, so we can just load them from there
  - the cert parameter is not needed in this implementation and whatever is passed to it should be ignored by it

**During validation:**

- only `getCert()` is used.  It must return a PEM string or buffer

> from crypto.createPublicKey(key) -> Creates and returns a new key object containing a public key. If key is a string or Buffer, format is assumed to be 'pem'

- keyInfo is used to get the public certificate from the keyInfo, but it can also be passed a custom public certificate 
  - if user created the instance with `new KeyInfo()` or `new KeyInfo(null)` - > `getCert(keyInfo)` extracts the cert from the first X509Certificate element
  - if user created the instance with `new KeyInfo(cert)`
      - > we should assume the user wants to validate the signature against the specific public certificate they passed to the constructor,
	  so getCert should return this.cert[0], regardless of if it was passed a parameter or not.
	  
The problem is that in this case, the HMAC signature will not work, because it doesn't use PEM keys, and that is why the tests were failing.
So what is the solution to that? Probably drop support for HMAC, or have a completely different workflow for using it.
HMAC is a symmetric mechanism where the same key is used to generate and verify the signature,
and thus doesn't fit in the standard public key infrastructure that XMLDSIG is designed around.

Reply:
Honestly, I thought about this a lot too. It seems to me that this entire thing should be static. If they call `getKeyInfoContents()`, they can pass in the certs they want. If the call `getCert()`, they can optionally specify a fallback public or private key (since a public key can be derived from a private key) that will be used if one can't be extracted from the `KeyInfo` block provided. That would solve lots of our problems and it would push the entire concept of `FileKeyInfo` out the window as file handling would be the consumer's problem (as it probably should be).

Thoughts?

Reply:
> Honestly, I thought about this a lot too. It seems to me that this entire thing should be static. If they call `getKeyInfoContents()`, they can pass in the certs they want. If the call `getCert()`, they can optionally specify a fallback public or private key (since a public key can be derived from a private key) that will be used if one can't be extracted from the `KeyInfo` block provided. That would solve lots of our problems and it would push the entire concept of `FileKeyInfo` out the window as file handling would be the consumer's problem (as it probably should be).
> 
> Thoughts?

I'm tapped out for thoughts currently, I don't think I have anything more to add than what I wrote above.
In any case, the HMAC thing needs dealing with as outlined in my comment. Those should not be used with KeyInfoProvider.

Reply:
@LoneRifle , perhaps you have fresh thoughts. We've known this part of the code is broken for a long time and it would be nice to fix it correctly now.

Reply:
A lot has been discussed though. I've never understood why a class was ever needed to hold key info. It's not like the class state (ie, the cert) is ever manipulated, or the helper methods produce an output that meaningfully combines class state with inputs.

I'm likewise inclined to break the class and turn it into a helper module, or if we can afford to, collapse its methods into SignedXml. We then manipulate certs solely in-memory, and leave it to the caller to figure out how to provide them, ie as-is, reading a file, or even async I/O.

Reply:
I'm guessing it was made that way so it is easy for the user to make their own implementation of the key provider.

Reply:
While true, thus far, the only practical example I've seen is from your PR comments in #283.

Maybe what might be best is a property specified by SignatureAlgorithm implementations that states whether an algorithm employs either public-key cryptography or a secret. SignatureAlgorithms that state they use shared secrets would then go on a separate code path from those that don't, as you outlined above.

Thoughts?

Reply:
> While true, thus far, the only practical example I've seen is from your PR comments in https://github.com/node-saml/xml-crypto/pull/283.

I'm not sure I follow what you are saying. I think in most cases users are using a custom `keyInfoProvider`. It's also documented [here](https://github.com/node-saml/xml-crypto#customizing-algorithms)

> Maybe what might be best is a property specified by SignatureAlgorithm implementations that states whether an algorithm employs either public-key cryptography or a secret. SignatureAlgorithms that state they use shared secrets would then go on a separate code path from those that don't, as you outlined above.
> 
> The package user will also be instructed to specify their secret explicitly as the private key in SignedXml, to drive home the point that the secret is meant to be kept as one. For schemes that employ public-key cryptography, both private and public keys will be specified.

That is a possible solution, but I think this conversation is better suited for issue #303
I don't see it as directly related to this PR.

Reply:
Nothing further to add at this point =)

Reply:
> I don't see any tests where KeyInfo or FileKeyInfo is used during the signature process. There are only tests using FileKeyInfo during validation process.

I've corrected your test and adjusted the code for `KeyInfo`. It now works and we have some coverage. Thank you for that contribution.

Reply:
Can you explain why the public cert needs to be set twice?

Reply:
The public cert needs to be set twice, because we don't use one of them. The public cert set in the ctor doesn't do anything when building the `<X509Certificate />` element, the only thing that does anything is the `publicCert` property. That is one of the reasons that I suggest that we just get rid of the ctor and make this entirely static or just remove the entire class and put two properties on `SignXml` for the two functions we need, with defaults, so they can override as needed.

Reply:
This overlooks the fact that there are other kinds of KeyInfo however, as @djaqua points out, and as @shunkica has shown in an example implementation of KeyInfoProvider. Extending SignXml to change the key provider might not be a good idea given that signing and verifying otherwise remains largely the same. That said, I'm okay for this to be done in the interim, while we figure out how to rebuild the KeyInfoProvider and/or SignXml's use of it.

After that, in the long run, SignXml can delegate to KeyInfoProvider or its successor once again.